Attributes

wojpok opened the pull request:
So far attributes may precede any `let` statement and may consist of multiple lists of identifiers separated by comma.
I am not really sure what to do with them at the moment, thus all attributes are dropped past de-sugar stage.

dominik-muc commented:
Do we only want attributes with `let`? I can see advantage of annotating a module or data-type.

However, this depends on how good do we want to implement it for now, thus what is really needed for test library (or maybe we can use it somewhere in lexer/parser generators?). It can be changed later on to support other constructions.

Also it might be a good idea to be consistent with list definitions, so maybe consider this syntax:

```
@{attr1}

@{[attr1, attr2]}

# Or to make things easier to read just list them one after another:

@{attr1}
@{attr2}
let _ = ...
```

Another question is whether we want attributes to take extra arguments. Should they behave like functions?
```
# This syntax might be useful for passing additional information to the compiler
@{attr3 "some param" 42}
let _ = ...
```

wojpok commented:
I have added attributes resolving to functions as discussed. As of now they are implemented over Surface Language, not Raw Language and honestly I am not really sure if that is the right way. Anyway, tests are properly toggleable from CLI

wojpok commented:
Since last meeting I have added new features to this pull request:
- Tests can be tagged and selected via command line using globs as discussed
- `abstr` attribute have been implemented
- Record are desugarred to ADT with tag `record`. Attribute resolver later adds missing record selectors
- Removed visibility from Raw language
- Improved attributes error handling

There is still work to be done, specifically test 89 fails

wojpok commented on src/DblParser/YaccParser.mly:
```diff
@@ -43,8 +43,9 @@ let make data =
-  if is_rec then make (DRec(false, [def])) else def
+  if is_rec then DRec([make ([], data)]) else data 
+
+let append xs ys = xs @ ys
```
I was worried that Yacc wont accept operators, but it turn out, everything is fine. This function is not needed

ppolesiuk commented on src/DblParser/Attributes.ml:
```diff
@@ -0,0 +1,304 @@
+    (attrs, dd :: sels)
+  | (xs, _) -> 
+    Error.fatal 
+      (Error.attribute_argument_arity_mismatch args.pos 0 (List.length xs))
```
The convention is that errors defined in `Error` module are programmer's faults, while internal errors are reported by `InterpLib.InternalError.report`. If the code is really unreachable, it should fall into the latter category.
